sys/xtimer/xtimer.c

```diff
@@ -212,41 +250,27 @@ static void _mutex_timeout(void *arg)
     unsigned int irqstate = irq_disable();
 
     mutex_thread_t *mt = (mutex_thread_t *)arg;
-
-    if (mt->mutex->queue.next != MUTEX_LOCKED &&
-        mt->mutex->queue.next != NULL) {
-        mt->timeout = 1;
-        list_node_t *node = list_remove(&mt->mutex->queue,
-                                        (list_node_t *)&mt->thread->rq_entry);
-
-        /* if thread was removed from the list */
-        if (node != NULL) {
-            if (mt->mutex->queue.next == NULL) {
-                mt->mutex->queue.next = MUTEX_LOCKED;
-            }
-            sched_set_status(mt->thread, STATUS_PENDING);
-            irq_restore(irqstate);
-            sched_switch(mt->thread->priority);
-            return;
-        }
-    }
+    mt->blocking = 0;
+    _mutex_remove_thread_from_waiting_queue(mt->mutex, mt->thread, &mt->got_unlocked);
     irq_restore(irqstate);
 }
 
 int xtimer_mutex_lock_timeout(mutex_t *mutex, uint64_t timeout)
 {
     xtimer_t t;
-    mutex_thread_t mt = { mutex, (thread_t *)sched_active_thread, 0 };
+    mutex_thread_t mt = { mutex, (thread_t *)sched_active_thread, .got_unlocked=0, .blocking=1 };
 
     if (timeout != 0) {
         t.callback = _mutex_timeout;
         t.arg = (void *)((mutex_thread_t *)&mt);
         xtimer_set64(&t, timeout);
     }
-
-    mutex_lock(mutex);
+    int ret = _mutex_lock(mutex, mt.blocking);
```

Comment:
handing the value of the volatile variable `blocking` to  `_mutex_lock` will not have the desired effect here. The timeout might deschedule you right before the `irq_disable` of `_mutex_lock` is called, leaving you with a blocking call where actually a non-blocking lock should happen. Is there any reason for not just simply doing a non-blocking lock before setting up the timeout timer? This way you would always try to lock the mutex and if the timeout then happens between setting up the timer and locking it, is actually desired behavior to not lock it, right?

Reply:
the timeout could happen before the mutex_lock but after the first non blocking mutex lock. Meaning there would be the same problem.
The problem with the irq_disable (interupt in function _mutex_lock but before irq_disable) will be fixed in the next PR. Tracking PR #11660
This PR fixes the problems with smaller timeouts.